src/dataflow-types/src/client/fat_client.rs

```diff
@@ -0,0 +1,274 @@
+    pub fn source_description_for(&self, id: GlobalId) -> Option<&crate::sources::SourceDesc> {
+        self.source_descriptions.get(&id).unwrap_or(&None).as_ref()
+    }
+
+    /// Returns the pair of `since` and `upper` for a maintained index, if it exists.
+    ///
+    /// The `since` frontier indicates that the maintained data are certainly valid for times greater
+    /// or equal to the frontier, but they may not be for other times. Attempting to create a dataflow
+    /// using this `id` with an `as_of` that is not at least `since` will result in an error.
+    ///
+    /// The `upper` frontier indicates that the data are reported available for all times not greater
+    /// or equal to the frontier. Dataflows with an `as_of` greater or equal to this frontier may not
+    /// immediately produce results.
+    pub fn index_since_upper_for(
+        &self,
+        id: GlobalId,
+    ) -> Option<(AntichainRef<Timestamp>, AntichainRef<Timestamp>)> {
+        self.compute_since_uppers.get(&id)
+    }
+
+    /// Returns the pair of `since` and `upper` for a source, if it exists.
+    ///
+    /// The `since` frontier indicates that the maintained data are certainly valid for times greater
+    /// or equal to the frontier, but they may not be for other times. Attempting to create a dataflow
+    /// using this `id` with an `as_of` that is not at least `since` will result in an error.
+    ///
+    /// The `upper` frontier indicates that the data are reported available for all times not greater
+    /// or equal to the frontier. Dataflows with an `as_of` greater or equal to this frontier may not
+    /// immediately produce results.
+    pub fn source_since_upper_for(
+        &self,
+        id: GlobalId,
+    ) -> Option<(AntichainRef<Timestamp>, AntichainRef<Timestamp>)> {
+        self.storage_since_uppers.get(&id)
+    }
+}
+
+#[derive(Default)]
+struct SinceUpperMap {
+    since_uppers:
+        std::collections::BTreeMap<GlobalId, (Antichain<Timestamp>, Antichain<Timestamp>)>,
+}
+
+impl SinceUpperMap {
+    fn insert(&mut self, id: GlobalId, since_upper: (Antichain<Timestamp>, Antichain<Timestamp>)) {
+        let prior = self.since_uppers.insert(id, since_upper);
+        assert!(prior.is_none());
+    }
+    fn get(&self, id: &GlobalId) -> Option<(AntichainRef<Timestamp>, AntichainRef<Timestamp>)> {
+        self.since_uppers
+            .get(id)
+            .map(|(since, upper)| (since.borrow(), upper.borrow()))
+    }
+    fn advance_since_for(&mut self, id: GlobalId, frontier: &Antichain<Timestamp>) {
+        if let Some((since, _upper)) = self.since_uppers.get_mut(&id) {
+            use differential_dataflow::lattice::Lattice;
+            since.join_assign(frontier);
+        } else {
+            // If we allow compaction before the item is created, pre-restrict the valid range.
+            // We start tracking `upper` at 0; correct this should that change (e.g. to `as_of`).
+            self.since_uppers
```

Comment:
This looks like it could interact poorly (panic) with `insert`, which asserts that prior was `None`. Is there some reason we know that `insert` will never be called after `advance_since_for`? It seems that `AllowIndexCompaction` or `AllowSourceCompaction` could reasonably be called before `CreateSources` or `CreateDataflows`, leading to a panic.

Reply:
Since this is auto-merging, is there now a requirement that coord never call allow compaction before create? That's fine, just making sure I understand.

Reply:
I thought I fixed this. If I did not, I will now.

Reply:
I think calling allow compaction before create is very weird, and might be a bug. I think in the future, `create` flavored command will need to come with a `GlobalId`, and if this id is not greater or equal to previously observed ids there will be an error. This is the mechanism that ensures no attempts at rebinding, while allowing us to maintain bounded state about the past.

I have a PR that makes this less crashy for the moment, but I think the behavior you describe should eventually result in some sort of error, yes.